saplings/circuits/src/components/forms/ProposeCircuitForm.js

```diff
@@ -14,18 +14,286 @@
+        if (!isSelected) {
+          return false;
+        }
+      }
+      return true;
+    }
+    if (node.displayName.toLowerCase().indexOf(lowerInput) > -1) {
+      return true;
+    }
+    return false;
+  });
+
+  return filteredNodes;
+};
+
+const nodesReducer = (state, action) => {
+  const errorMessage =
+    'At least two nodes must be part of a circuit. Please select a node.';
+
+  switch (action.type) {
+    case 'filter': {
+      const nodes = filterNodes(state, action.input);
+      const filteredNodes = {
+        nodes,
+        filteredBy: action.input
+      };
+      return { ...state, filteredNodes };
+    }
+    case 'addLocalNode': {
+      const localNode = action.node;
+      state.selectedNodes.push(localNode);
+      state.filteredNodes.nodes.sort((node1, node2) => {
+        if (node1.identity === localNode.identity) {
+          return -1;
+        }
+        if (node2.identity === localNode.identity) {
+          return 1;
+        }
+        return 0;
+      });
+      return { ...state };
+    }
+    case 'showSelectedOnly': {
+      const newState = state;
+      newState.showSelectedOnly = true;
+      const nodes = filterNodes(state, state.filteredNodes.filteredBy);
+      const filteredNodes = {
+        nodes,
+        filteredBy: state.filteredNodes.filteredBy
+      };
+      return { ...newState, filteredNodes };
+    }
+    case 'showAllNodes': {
+      const newState = state;
+      newState.showSelectedOnly = false;
+      const nodes = filterNodes(state, state.filteredNodes.filteredBy);
+      const filteredNodes = {
+        nodes,
+        filteredBy: state.filteredNodes.filteredBy
+      };
+      return { ...newState, filteredNodes };
+    }
+    case 'toggleSelect': {
+      const { node } = action;
+      let alreadySelected = false;
+
+      const selectedNodes = state.selectedNodes.filter(selectedNode => {
+        if (node.identity === selectedNode.identity) {
+          alreadySelected = true;
+          return false;
+        }
+        return true;
+      });
+
+      if (!alreadySelected) {
+        selectedNodes.push(node);
+      }
+
+      const nodes = filterNodes(state, state.filteredNodes.filteredBy);
+      const filteredNodes = {
+        nodes,
+        filteredBy: state.filteredNodes.filteredBy
+      };
+
+      let { error } = state;
+      if (selectedNodes.length >= 2) {
+        error = '';
+      } else {
+        error = errorMessage;
+      }
+
+      return { ...state, selectedNodes, filteredNodes, error };
+    }
+    case 'removeSelect': {
+      const { node } = action;
+      const selectedNodes = state.selectedNodes.filter(
+        item => item.identity !== node.identity
+      );
+
+      const nodes = filterNodes(state, state.filteredNodes.filteredBy);
+      const filteredNodes = {
+        nodes,
+        filteredBy: state.filteredNodes.filteredBy
+      };
+
+      let error = '';
+      if (selectedNodes.length < 2) {
+        error = errorMessage;
```

Comment:
This error message should probably be specific to the actual error

Reply:
updated